### DISCUSSION TITLE
colour.dominant_wavelength(xy, xy_n)

### DISCUSSION THREAD

**Discussion by fabio-ric-silva** [colour-science/colour#1249] [General]
### Question

Dears;
  I have read some documentation, but couldn't find how to get the value of xy_n.
  My sensor reads RGB, and I'm able of getting the XYZ value, and converting for CIA XY.
  
 But, how can I get the xy_n value?
  
Thanks in advance!

Sincerely;

**Comment by tjdcs** [comment]
Hi @fabio-ric-silva 

xy_n is the "white point" or the point in x,y space that has no dominant wavelength and it is selected by you, the user, based on your application. 

In CIE Colorimetry and in the traditional definition of dominant wavelength + purity, the equal energy spectrum is defined to be the reference point. Under this assumption / standard the `xy_n` point is (1/3, 1/3) 

However, in other applications and in research there may be other useful choices like D65, 7,000K or 15,000K. Unless you know you are specifically working in those fields and advanced research you should use the CIE standard which I believe is the default value for this function. 

Hope this helps!

**Reply by fabio-ric-silva** [reply]
OK, so, in my case:
  My device is an RGB sensor used for extracting the PEAK wavelength from a light source!
   I'm using the TCS34007FN, and using it for reading (blue light LEDs).
   https://ams.com/documents/20143/36005/TCS3400_DS000411_5-00.pdf/5e66e18f-b76c-3245-47a7-1813b423c846
   The CIA calculations, I have gotten from the website:

''' **STEP 1**
    Get the RGB values from your color object and convert them to be 
    between 0 and 1. So the RGB color (255, 0, 100) becomes (1.0, 0.0, 0.39)
    '''
    try:
        Clear   =   data[1] + data[2] + data[3] #(data[1] + data[2] + data[3])
        Red     =   data[1] / Clear   
        Green   =   data[2] / Clear    
        Blue    =   data[3] / Clear 
        print('R + G + B - STANDARD', Red * 255, Green * 255, Blue * 255, '\n' )    
    except: 
        Clear   =   0
        Red     =   0  
        Green   =   0  
        Blue    =   0        



    **''' STEP 2**
    Apply a gamma correction to the RGB values, which makes the color more vivid 
    and more the like the color displayed on the screen of your device. This gamma 
    correction is also applied to the screen of your computer or phone, thus we need 
    this to create the same color on the light as on screen. This is done by the following formulas:
     float red = (red > 0.04045f) ? pow((red + 0.055f) / (1.0f + 0.055f), 2.4f) : (red / 12.92f); 
     float green = (green > 0.04045f) ? pow((green + 0.055f) / (1.0f + 0.055f), 2.4f) : (green / 12.92f);
     float blue = (blue > 0.04045f) ? pow((blue + 0.055f) / (1.0f + 0.055f), 2.4f) : (blue / 12.92f);
    '''
    if Red > 0.04045:
         Red_Gamma   = pow((Red + 0.055) / (1.0 + 0.055), 2.4) 
    else:
         Red_Gamma   = Red / 12.92 

    if Green > 0.04045:
         Green_Gamma   = pow((Green + 0.055) / (1.0 + 0.055), 2.4) 
    else:
         Green_Gamma   = Green / 12.92

    if Blue > 0.04045:
         Blue_Gamma   = pow((Blue + 0.055)  / (1.0 + 0.055), 2.4) 
    else:
         Blue_Gamma   = Blue  / 12.92          
    
    '''  **STEP 3**
    Convert the RGB values to XYZ using the Wide RGB D65 conversion formula The formulas used:
    float X = red * 0.649926f + green * 0.103455f + blue * 0.197109f;
    float Y = red * 0.234327f + green * 0.743075f + blue * 0.022598f;
    float Z = red * 0.0000000f + green * 0.053077f + blue * 1.035763f;
    '''
    X_RGB = Red_Gamma * 0.649926  + Green_Gamma * 0.103455 + Blue_Gamma * 0.197109
    Y_RGB = Red_Gamma * 0.234327  + Green_Gamma * 0.743075 + Blue_Gamma * 0.022598
    Z_RGB = Red_Gamma * 0.0000000 + Green_Gamma * 0.053077 + Blue_Gamma * 1.035763

    ''' **STEP 4**
    Calculate the xy values from the XYZ values
    '''
    try:
        X_CIA = X_RGB / (X_RGB + Y_RGB + Z_RGB)
        Y_CIA = Y_RGB / (X_RGB + Y_RGB + Z_RGB)
        Z_CIA = 1 - X_CIA - Y_CIA
        #convert to CCT (temperature color)
        n    = (X_CIA - 0.3320) / (0.1858 - Y_CIA)
        #CCT  = (449.0 * ( n**3)) + (3525.0 * (n**2)) + (6823.3 * n) + 5520.33 
        CCT  = (437.0 * ( n**3)) + (3601.0 * (n**2)) + (6861.0 * n) + 5517.0 
        
    except:
        X_CIA = "FAIL"
        Y_CIA = "FAIL"
        Z_CIA = "FAIL"
        CCT   = "FAIL"



    print('R + G + B + GAMA CORRECT', Red_Gamma * 255, Green_Gamma * 255, Blue_Gamma * 255,'\n' )    
    print('Xcia, Ycia, Zcia: ', X_CIA, Y_CIA, Z_CIA,'\n')
    print('Correlated Color Temperature: ', CCT, '\n')

 At this point, I really don't know if I would have to use CIA our just RGB for reading just the peak wavelength from a light led source!

    **_#wavelength = [X_CIA,Y_CIA]
    #xyclear = [1/3,1/3]
    #t = colour.dominant_wavelength(wavelength, xyclear)._**

  Im asking it, because I tryed using a website with the CIAx and CIAy results for getting the dominant wavelength ad the result was absolutely perfect, matching the results of tests carried out in laboratories.

 The website that i have used is therebelow:

 https://luminus-cie1931-demo.anvil.app/

Thank you so much by your help DEAR!

Sincerely;

Fabio!
